### ISSUE TITLE
missing binary files in 2.0.26-lto

### ISSUE THREAD

**Issue by waterlike86** [emscripten-core/emsdk#866]
2.0.26 has 45 files in upstream/bin

```
FileCheck  clang-offload-wrapper  llc      llvm-cov        llvm-dwp   llvm-objcopy   llvm-rc       llvm-strip       wasm-ctor-eval            wasm-metadce  wasm-split      wasm32-wasi-clang
clang      diagtool               lld      llvm-cxxfilt    llvm-link  llvm-objdump   llvm-readobj  llvm-symbolizer  wasm-dis                  wasm-opt      wasm2js         wasm32-wasi-clang++
clang++    git-clang-format       llvm-ar  llvm-dis        llvm-mc    llvm-profdata  llvm-size     opt              wasm-emscripten-finalize  wasm-reduce   wasm32-clang
clang-13   hmaptool               llvm-as  llvm-dwarfdump  llvm-nm    llvm-ranlib    llvm-strings  wasm-as          wasm-ld                   wasm-shell    wasm32-clang++
```

2.0.26-LTO has 34 files
```
clang     llc             llvm-ar         llvm-dwp      llvm-objdump  llvm-size        opt             wasm-dis                  wasm-metadce  wasm-shell  wasm32-clang       wasm32-wasi-clang++
clang++   lld             llvm-cxxfilt    llvm-nm       llvm-ranlib   llvm-strings     wasm-as         wasm-emscripten-finalize  wasm-opt      wasm-split  wasm32-clang++
clang-13  llvm-addr2line  llvm-dwarfdump  llvm-objcopy  llvm-readobj  llvm-symbolizer  wasm-ctor-eval  wasm-ld                   wasm-reduce   wasm2js     wasm32-wasi-clang
```
is this expected? was trying but llvm-link is missing from LTO and my local LTO build fails

**Comment by waterlike86** [comment]
@dschuff maybe you might know since i see that you have been updating the LTO builds, could you please help here?

**Comment by sbc100** [comment]
I think this is expected.  We trimmed binaries that are not designed for distribution (e.g. FileCheck).

**Comment by sbc100** [comment]
I guess llvm-link should not have been removed.

BTW do you really need llvm-link?  Are you relying on emcc's `-shared` flag to merge object files using llvm-link?   Can you not use llvm-ar to create archive files instead?

**Comment by waterlike86** [comment]
we are using LTO in our builds when creating libs and the final wasm target. and it seems that this causes emscripten to use llvm-link instead of LLD. probably also because we are using .bc as the extension for our libs
https://github.com/emscripten-core/emscripten/blob/37bf094183758fef2dadd9350c86e1ecfd70a89b/tools/building.py#L342

**Comment by sbc100** [comment]
If you can, I would recommend switching to using static libraries (i.e. use `emar` to create them rather than emcc).

**Comment by dschuff** [comment]
I didn't know there was any way to invoke emscripten anymore that caused it to link bitcode. 
@waterlike86 can you give an example use case with command lines? 
@sbc100 do you know of any use cases where we actually need `llvm-link` and can't use lld?

**Comment by sbc100** [comment]
Yes, if we are creating a combined object files (using `-r` or `-shared`) and `-flto` is used then we need to link bitcode together.   

I'd love to remove this feature one day..

**Comment by sbc100** [comment]
Remember that `-shared` today does not create a runtime shared library (i.e. a side module) but just a big object file.

**Comment by dschuff** [comment]
1) `-r` with `-flto` seems like an extremely niche use case. Since the result is still just a regular object file, are there any use cases where you couldn't just run the LTO at the time you do the final link (and it seems like it would end up with better code in that case anyway, since it can internalize everything)

2) once we have `-shared` in the state where we want it, we can let lld do LTO for shared libraries, right? Then we wouldn't need llvm-link?

**Comment by sbc100** [comment]
> 1. `-r` with `-flto` seems like an extremely niche use case. Since the result is still just a regular object file, are there any use cases where you couldn't just run the LTO at the time you do the final link (and it seems like it would end up with better code in that case anyway, since it can internalize everything)

Sure it niche, but some folks (like the OP) are doing this.    My recommendation is always to just use `emar` archives instead.  (is that what you mean by "just run the LTO at the time you do the final link"?)

> 2. once we have `-shared` in the state where we want it, we can let lld do LTO for shared libraries, right? Then we wouldn't need llvm-link?

**Comment by dschuff** [comment]
Oh, I was thinking about when to run the cross-module optimizations. I'm saying there's no point in doing that when creating the joined bitcode file (rather than at final link time) because you can't internalize. But I guess even if you planned to do LTO at final link time, then your intermediate intputs would still be bitcode. 

And yeah, using .a archives with a bunch of object files instead of a single one (created with `-r`) is what most people want. There is a use case for incremental linking with `-r` although it's often about making the final link faster, which would be defeated by LTO. 
Technically if we remove llvm-link it would prevent incremental linking of bitcode files.  I guess people are doing that, I just don't know what the advantages are over doing multi-object archive files with LTO at link time.

**Comment by dschuff** [comment]
But for now I can put llvm-link back in the LTO build. If we are going to break people, we should do it on purpose rather than by accident :)

**Comment by sbc100** [comment]
> Oh, I was thinking about when to run the cross-module optimizations. I'm saying there's no point in doing that when creating the joined bitcode file (rather than at final link time) because you can't internalize. But I guess even if you planned to do LTO at final link time, then your intermediate intputs would still be bitcode.
> 
> And yeah, using .a archives with a bunch of object files instead of a single one (created with `-r`) is what most people want. There is a use case for incremental linking with `-r` although it's often about making the final link faster, which would be defeated by LTO.
> Technically if we remove llvm-link it would prevent incremental linking of bitcode files. I guess people are doing that, I just don't know what the advantages are over doing multi-object archive files with LTO at link time.

We also can't get rid of the fake support for `-shared` just yet..

**Comment by waterlike86** [comment]
>     1. `-r` with `-flto` seems like an extremely niche use case. Since the result is still just a regular object file, are there any use cases where you couldn't just run the LTO at the time you do the final link (and it seems like it would end up with better code in that case anyway, since it can internalize everything)

we dont use -r, just -flto, we are not using emcc directly but using cmake for build.
for example for compilation to objects we use flags
  FLAGS = -fPIC -flto -Oz
then during linking we use flags below to create a .bc file. 
  LINK_FLAGS = -flto -fPIC -s WASM_BIGINT=1 -Oz
  
  for legacy reasons we have a final .bc file, which basically combines all the dependent libraries into 1, before we turn that into a wasm. I expect that this would actually churn out bitcode libraries, @dschuff is this not the case?
  
  
  @sbc100 just to clarify 
  > My recommendation is always to just use emar archives instead
  
  does this mean that the extension of the libraries matter? i see that theres special handling for .bc files and it never tries to use emar if the extension is .bc

**Comment by sbc100** [comment]
> > ```
> > 1. `-r` with `-flto` seems like an extremely niche use case. Since the result is still just a regular object file, are there any use cases where you couldn't just run the LTO at the time you do the final link (and it seems like it would end up with better code in that case anyway, since it can internalize everything)
> > ```
> 
> we dont use -r, just -flto, we are not using emcc directly but using cmake for build.
> for example for compilation to objects we use flags
> FLAGS = -fPIC -flto -Oz
> then during linking we use flags below to create a .bc file.
> LINK_FLAGS = -flto -fPIC -s WASM_BIGINT=1 -Oz
> 
> for legacy reasons we have a final .bc file, which basically combines all the dependent libraries into 1, before we turn that into a wasm. I expect that this would actually churn out bitcode libraries, @dschuff is this not the case?
> 
> @sbc100 just to clarify
> 
> > My recommendation is always to just use emar archives instead
> 
> does this mean that the extension of the libraries matter? i see that theres special handling for .bc files and it never tries to use emar if the extension is .bc

No the extension should not matter.   emcc will look at the file contents not the extension.

How are you creating your "final .bc file", are you calling `llvm-link` directly?  With emcc I think the only way to create a combined bc file is to use `-r` or `-shared`.

**Comment by sbc100** [comment]
While we get `llvm-link` added back to the sdk would you mind seeing if you can switch away from using `-r` and either link to a final `.wasm` file, or use and `emar` archive for your intermediate step?

Alternatively you could drop the `-flto` from your LINK_FLAGS which would mean the emcc would perform a link to and object file rather than a `.bc` file during that step.   Oddly by removing the `-flto` flag you would actually be performing LTO at that step.

**Comment by sbc100** [comment]
@dschuff how were we able to pass all the tests if `wasm-link` was not part of the bundle?  Are we testing against a different set of binaries the ones that end up in the final tarball?

**Comment by waterlike86** [comment]
>While we get llvm-link added back to the sdk would you mind seeing if you can switch away from using -r and either link to a final .wasm file, or use and emar archive for your intermediate step?

yes i will try that. for now i continued using the non LTO version of the sdk.

>Alternatively you could drop the -flto from your LINK_FLAGS which would mean the emcc would perform a link to and object file rather than a .bc file during that step. Oddly by removing the -flto flag you would actually be performing LTO at that step.

this is rather confusing. i would expect that dropping -flto from the compilation flags should be the way how does dropping it from the LINK_FLAGS cause LTO to be performed?

**Comment by sbc100** [comment]
> > While we get llvm-link added back to the sdk would you mind seeing if you can switch away from using -r and either link to a final .wasm file, or use and emar archive for your intermediate step?
> 
> yes i will try that. for now i continued using the non LTO version of the sdk.
> 
> > Alternatively you could drop the -flto from your LINK_FLAGS which would mean the emcc would perform a link to and object file rather than a .bc file during that step. Oddly by removing the -flto flag you would actually be performing LTO at that step.
> 
> this is rather confusing. i would expect that dropping -flto from the compilation flags should be the way how, does dropping it from the LINK_FLAGS cause LTO to be performed?

Dropping it from LINK_FLAGS in this case will caus `-r` to call wasm-ld rather than wasm-link to perform the linking and wasm-ld will perform the LTO step at this point, producing your single object file (that thing you call `.bc` today but should probably be better called `.o`).   Without `-flto` this will be a normal wasm object file represnting the LTO'd link of all your input objects.  

Does that make sense.   Here is the code that implements the `-r` flag:

https://github.com/emscripten-core/emscripten/blob/58647ae5b0fa8af834871c26b3839f5bd84dd3d8/tools/building.py#L336-L342

**Comment by dschuff** [comment]
> @dschuff how were we able to pass all the tests if `wasm-link` was not part of the bundle? Are we testing against a different set of binaries the ones that end up in the final tarball?

Emscripten CI uses the latest build uploaded from emscripten-releases. When the latest build is the LTO build, it will use that. But otherwise it uses the non-LTO builds (which is most of the time). The emscripten-releases CI does run tests on what it builds as well, but we don't run quite as many tests as Emscripten CI. Which test suite catches this exactly?

**Comment by gl84** [comment]
I use a similar approach as @waterlike86 when building wasm files...

I think removal of llvm-link (or removing the linking of bitcode libraries in general) is not a good idea, at least not for how i use emscripten today.

As compiling with emscripten can be quite time consuming I compile all libraries I use to bitcode (also with -flto=thin). Further I use opt to reduce the number of symbols to only the ones I really need.

This allows me to use a single build for both static and shared libraries and allows different final wasm-builds:
   - reduced size builds with global DCE (single wasm file that contains multiple bitcode libraries)
   - incremental development builds (one wasm file per library)

If support for linking bitcode libraries is removed I'm stuck with either:
   - huge static libraries with many unwanted symbols
   - multiple shared libraries without global DCE

Just my 2 cents...

**Comment by sbc100** [comment]
@gl84 we are not planning on actually removing llvm-link, is was removed by mistake and will be re-added.

However, have you tried building with wasm object files?  If you want to fast build I believe you will get the best results from building each source directly to wasm object files (i.e. don't pass -flto at all).  This way you can compile in parallel and link very quickly.   Having large static libraries (created using emar) made of up wasm object files should be very fast since the linker will only pick out the symbols it need and will not have do do any compilation.   This is how you should be able to achieve fast builds.

Of course if you want/need LTO then the build will be a lot slower but that is to be expected. 

Can I ask how you are using llvm-link exactly?  Are you using `emcc -r`, or `emcc -shared` or calling `llvm-link` yourself?

**Comment by sbc100** [comment]
The reason I ask is that I don't think either LTO or non-LTO builds should ultimately require of benefit from wasm-link.  You should be able to get good reasults form static librarys (created by amar) for both LTO builds and non-LTO builds.

**Comment by gl84** [comment]
@sbc Cool to hear that!

I checked different build types (LTO, non-LTO, static libraries) about a year ago and found that for me calling llvm-link directly is the fastest/best solution.

In my case the linking of hundreds of separate test cases was at that time super slow (even without LTO). As I only use a small subset of multiple 30mb+ static libraries I decided to link all of them together and remove what I don't need/use...

But since emscripten is moving so fast this maybe isn't necessary anymore...

**Comment by sbc100** [comment]
Yes, in the past we didn't have wasm object files so all linking was slow.  These days, especially if you want to link lots of small programs, you should avoid LTO and use wasm object files.  This will give you by far the fastest link times. 

It shouldn't matter how much unused object files or libraries you have in your link since wasm-ld will only include what is needed on-demand and doesn't need to do any compilation at all so should be blazing fast.

**Comment by waterlike86** [comment]
another issue which i had was trying to link the libraries individually to the executable target. but the scripts doesnt seem to accept -Wl,--whole-archive for link libs it would eventually be stripped away before being passed to wasm-ld.

**Comment by sbc100** [comment]
I imagine you probably don't want `-r` there.   `-r` is for creating relocatable object files (inputs for the linker), not run-able things.

**Comment by sbc100** [comment]
Its not very easy to tell because LTO will be run if any of the input files are LTO objects (i.e. bitcode).   

Normally one sets up the build system to create LTO object files in LTO configuration and non-LTO object files in non-LTO configuration.    I guess you could run the linker with `--verbose` .. it might print information about if/when LTO is run.
